web-api/src/personalization/filter.rs

```diff
@@ -0,0 +1,201 @@
+// Copyright 2023 Xayn AG
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, version 3.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use std::fmt;
+
+use serde::{
+    de::{Error, MapAccess, Unexpected, Visitor},
+    Deserialize,
+    Deserializer,
+};
+
+use crate::models::{DocumentProperty, DocumentPropertyId};
+
+#[derive(Clone, Copy, Debug, Deserialize, Eq, Hash, PartialEq)]
+pub(crate) enum CompareOp {
+    #[serde(rename = "$eq")]
+    Eq,
+}
+
+#[derive(Debug, PartialEq)]
+struct CompareWith {
+    operation: CompareOp,
+    value: DocumentProperty,
+}
+
+impl<'de> Deserialize<'de> for CompareWith {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        struct CompareWithVisitor;
+
+        impl<'de> Visitor<'de> for CompareWithVisitor {
+            type Value = CompareWith;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+                formatter.write_str("a json object")
+            }
+
+            fn visit_map<A>(self, mut map: A) -> Result<Self::Value, A::Error>
+            where
+                A: MapAccess<'de>,
+            {
+                const EXPECTED: &&str = &"exactly one right comparison argument";
+                let Some((operation, value)) = map.next_entry::<_, DocumentProperty>()? else {
+                    return Err(A::Error::invalid_length(0, EXPECTED));
+                };
+                if map.next_entry::<CompareOp, DocumentProperty>()?.is_some() {
+                    return Err(A::Error::invalid_length(
+                        2 + map.size_hint().unwrap_or_default(),
+                        EXPECTED,
+                    ));
+                }
+                match operation {
+                    CompareOp::Eq => {
+                        if !value.is_string() {
+                            return Err(A::Error::invalid_type(
+                                Unexpected::Other("no string property"),
+                                &"a string property",
+                            ));
+                        }
+                    }
+                }
+
+                Ok(CompareWith { operation, value })
+            }
+        }
+
+        deserializer.deserialize_map(CompareWithVisitor)
```

Comment:
:question: I'm not sure if I'm missing something but why do it so complicated.

Couldn't we have used something like:

```rust
#[derive(Deserialize)]
struct LeafFilter {
    property: DocumentPropertyId,
    operator: LeafFilterOperator,
}

#[derive(Serialize)]
enum LeafFilterOperator {
  #[serde(rename="$eq")]
  Eq(AnyPropertyValue),
  //... Ne, Lt, Lte, In, etc. 
  #[serde(rename="$lt")]
  Lt(OrderedPropertyValue),
  #[serde(rename="$in")]
  In(Vec<DocumentPropertyString>),
}
```

or if you want to have better handcrafted error messages but in turn also more 
post evaluation by hand:

```rust
type LeafFilterOperator = HashMap<LeafOperator, Value>; 
``` 

sure for `$and`/`$or` this will still require some usage of `#[serde(untagged)]` or even
manual code, but it should be in general easier to implement this way.

Reply:
this was pretty much my line of thought, i went from enum to hashmap to singlevaluehashmap. the main difference is that i flattened the `SingleValueMap<DocumentPropertyId, SingleValueMap<LeafOperator, Value>>` which is essentially the `Compare` and the inner `SingleValueMap<LeafOperator, Value>` is the `CompareWith` and the `LeafOperator` is the `CompareOp`.
with `Compare` and `CompareWith` we can get better error message where the input was invalid, with just a generic `SingleValueMap` this would be more difficult. if that's not needed, than i can also make it more generic, but it seemed good to have.